s3torchbenchmarking/src/s3torchbenchmarking/dcp/benchmark.py

```diff
@@ -0,0 +1,154 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  // SPDX-License-Identifier: BSD
+
+import logging
+import os
+import random
+import string
+from multiprocessing.queues import Queue
+from pathlib import Path
+from time import perf_counter
+from typing import List
+
+import hydra
+import numpy as np
+import torch
+import torch.distributed as dist
+import torch.distributed.checkpoint as dcp
+from omegaconf import DictConfig
+from torch import multiprocessing
+from torch.distributed.checkpoint import FileSystemWriter
+from torch.nn import Module
+from torch.nn.parallel import DistributedDataParallel
+
+from s3torchbenchmarking.benchmark_utils import ResourceMonitor
+from s3torchbenchmarking.dcp.distribution import Distribution
+from s3torchbenchmarking.dcp.models import get_benchmark_model
+from s3torchbenchmarking.dcp.results import save_results
+from s3torchconnector.dcp import S3StorageWriter
+
+logger = logging.getLogger(__name__)
+
+
+@hydra.main(version_base=None, config_path=".", config_name="config")
+def run_benchmark(cfg: DictConfig):
+    """DCP benchmark entry point."""
+    benchmark_model = get_benchmark_model(cfg.model)
+    world_size = validate_world_size(cfg.world_size)
+
+    # For every run, use a randomized suffix (for either local disk or S3).
+    suffix = "".join(random.choices(string.ascii_letters, k=7))
+    storage_writer = get_writer(cfg, suffix)
+
+    manager = multiprocessing.Manager()
+    save_timestamps: Queue[float] = manager.Queue()
+
+    processing_times = Distribution()
+    ptp_save_times = Distribution()  # "ptp" == "peak to peak", or "diff(max, min)"
+
+    with ResourceMonitor() as monitor:
+        start_time = perf_counter()
+        multiprocessing.spawn(
+            run,
+            (
+                cfg.backend,
+                cfg.epochs,
+                benchmark_model.model,
+                world_size,
+                storage_writer,
+                save_timestamps,
+            ),
+            nprocs=world_size,
+            join=True,
+        )
+        end_time = perf_counter()
+        processing_times.append(end_time - start_time)
+
+        collector: List[float] = []
+        while not save_timestamps.empty():
+            collector.append(save_timestamps.get())
+        ptp_save_times.append(np.ptp(collector))
+
+    save_results(
+        cfg,
+        benchmark_model,
+        ptp_save_times=ptp_save_times,
+        processing_times=processing_times,
+        monitor=monitor,
+    )
+
+
+def validate_world_size(world_size: int) -> int:
+    """Enforce `world_size` to be within the current node's capacity.
+
+    FIXME: only works when the backend is "nccl"; what about "gloo" and CPUs?
+    """
+    device_count = torch.cuda.device_count()
+    if world_size > device_count:
```

Comment:
I don't think we should restrict this, world size can be greater than number of gpus, there is no limitation like this.

Reply:
Got it, lifting that up